ProtoArray: Best node is not viable for head

I upgraded to Teku v21.3.2 at Mar 25 23:27:59. On Mar 26 I noticed an unusually high number of missed attestations as well as Exceptions. Rolling back to v21.3.0 seems to have resolved the issues.

This is an example Exception:

java.util.concurrent.CompletionException: java.lang.RuntimeException: ProtoArray: Best node is not viable for head
        at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:367) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:376) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1019) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:142) ~[teku-infrastructure-async-2>
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2251) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:501) ~[teku-infrastructure-async-21.3.2.jar:21>
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:29) ~[teku-infrastructure-async-21.3.2.jar:21.>
        at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:139) ~[teku-infrastructure-async-21.3.2.jar>
        at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateTo(SafeFuture.java:236) ~[teku-infrastructure-async-21.3.2.jar:21.>
        at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRu>
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.RuntimeException: ProtoArray: Best node is not viable for head
        at tech.pegasys.teku.protoarray.ProtoArray.findHead(ProtoArray.java:175) ~[teku-protoarray-21.3.2.jar:21.3.2]
        at tech.pegasys.teku.protoarray.ForkChoiceStrategy.findHead(ForkChoiceStrategy.java:94) ~[teku-protoarray-21.3.2.jar:21.3.2]
        at tech.pegasys.teku.statetransition.forkchoice.ForkChoice.findNewChainHead(ForkChoice.java:254) ~[teku-ethereum-statetransition>
        at tech.pegasys.teku.statetransition.forkchoice.ForkChoice.updateForkChoiceForImportedBlock(ForkChoice.java:225) ~[teku-ethereum>
        at tech.pegasys.teku.statetransition.forkchoice.ForkChoice.lambda$onBlock$4(ForkChoice.java:197) ~[teku-ethereum-statetransition>
        at tech.pegasys.teku.infrastructure.async.eventthread.AsyncRunnerEventThread.recordEventThreadIdAndExecute(AsyncRunnerEventThrea>
        at tech.pegasys.teku.infrastructure.async.eventthread.AsyncRunnerEventThread.lambda$doExecute$1(AsyncRunnerEventThread.java:92) >
        at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:78) ~[teku-infrastructure-async-21.3.2.jar:21.3.2]
        at tech.pegasys.teku.infrastructure.async.AsyncRunner.lambda$runAsync$2(AsyncRunner.java:38) ~[teku-infrastructure-async-21.3.2.>
        at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:70) ~[teku-infrastructure-async-21.3.2.jar:21.3.2]
        ... 4 more

* Teku version: teku/v21.3.2/linux-x86_64/-ubuntu-openjdk64bitservervm-java-11
* Java version: OpenJDK 64-Bit Server VM (build 11.0.10+9-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)
* Distro: Ubuntu 20.04
* Kernel: 5.8.0-48-generic #54~20.04.1-Ubuntu SMP Sat Mar 20 13:40:25 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,3 +26,4 @@ For information on changes in released versions of Teku, see the [releases page]
 ### Bug Fixes
 - Fixed issue in discv5 where nonce was incorrectly reused.
 - Block events now only return the slot and root, rather than the entire signed block.
+- Fixed `ProtoArray: Best node is not viable for head` error.

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java
@@ -247,9 +247,6 @@ public class ForkChoice {
       return new SlotAndBlockRoot(block.getSlot(), block.getRoot());
     }
 
-    // TODO: Can we optimise and avoid running findHead if the justified checkpoint changes?
-    // This can be the only block we have with the new justified checkpoint so must be the head
-
     // Otherwise, use fork choice to find the new chain head as if this block is on time the
     // proposer weighting may cause us to reorg.
     // During sync, this may be noticeably slower than just comparing the chain head due to the way

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java
@@ -247,11 +247,16 @@ public class ForkChoice {
       return new SlotAndBlockRoot(block.getSlot(), block.getRoot());
     }
 
+    // TODO: Can we optimise and avoid running findHead if the justified checkpoint changes?
+    // This can be the only block we have with the new justified checkpoint so must be the head
+
     // Otherwise, use fork choice to find the new chain head as if this block is on time the
     // proposer weighting may cause us to reorg.
     // During sync, this may be noticeably slower than just comparing the chain head due to the way
     // ProtoArray skips updating all ancestors when adding a new block but it's cheap when in sync.
-    return forkChoiceStrategy.findHead(recentChainData.getJustifiedCheckpoint().orElseThrow());
+    final Checkpoint justifiedCheckpoint = recentChainData.getJustifiedCheckpoint().orElseThrow();
+    final Checkpoint finalizedCheckpoint = recentChainData.getFinalizedCheckpoint().orElseThrow();
+    return forkChoiceStrategy.findHead(justifiedCheckpoint, finalizedCheckpoint);
   }
 
   public SafeFuture<AttestationProcessingResult> onAttestation(

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceTest.java
@@ -286,7 +286,7 @@ class ForkChoiceTest {
 
     // Update ProtoArray to avoid the special case of considering the anchor
     // epoch as allowing all nodes to be a valid head.
-    forkChoice.processHead(epoch4StartSlot);
+    processHead(epoch4StartSlot);
 
     prepEpochForJustification(chainUpdater, epoch4StartSlot);
 

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceTest.java
@@ -44,6 +44,7 @@ import tech.pegasys.teku.spec.datastructures.util.AttestationProcessingResult;
 import tech.pegasys.teku.spec.logic.common.statetransition.results.BlockImportResult;
 import tech.pegasys.teku.spec.util.DataStructureUtil;
 import tech.pegasys.teku.storage.api.TrackingChainHeadChannel.ReorgEvent;
+import tech.pegasys.teku.storage.client.ChainUpdater;
 import tech.pegasys.teku.storage.client.RecentChainData;
 import tech.pegasys.teku.storage.server.StateStorageMode;
 import tech.pegasys.teku.storage.storageSystem.InMemoryStorageSystemBuilder;
@@ -264,6 +265,81 @@ class ForkChoiceTest {
     // And we should be able to apply the new weightings without making the fork block's weight
     // negative
     assertDoesNotThrow(() -> forkChoice.processHead(updatedAttestationSlot));
+  }
+
+  @Test
+  void onBlock_shouldHandleNonCanonicalBlockThatUpdatesJustifiedCheckpoint() {
+    // If the new block is not the child of the current head block we use `ProtoArray.findHead`
+    // to check if it should become the new head.  If importing that block caused the justified
+    // checkpoint to be updated, then the justified epoch in ProtoArray won't match the justified
+    // epoch of the new head so it considers the head invalid.  Normally that update is done when
+    // applying pending votes.
+
+    final ChainUpdater chainUpdater = storageSystem.chainUpdater();
+    final UInt64 epoch4StartSlot = spec.computeStartSlotAtEpoch(UInt64.valueOf(4));
+
+    // Set the time to be the start of epoch 3 so all the blocks we need are valid
+    chainUpdater.setTime(
+        spec.getSlotStartTime(epoch4StartSlot, genesis.getState().getGenesis_time()));
+
+    justifyEpoch(chainUpdater, 2);
+
+    // Update ProtoArray to avoid the special case of considering the anchor
+    // epoch as allowing all nodes to be a valid head.
+    forkChoice.processHead(epoch4StartSlot);
+
+    prepEpochForJustification(chainUpdater, epoch4StartSlot);
+
+    // Switch head to a different fork so the next block has to use findHead
+    chainUpdater.updateBestBlock(chainBuilder.fork().generateNextBlock());
+
+    final SignedBlockAndState epoch4Block = chainBuilder.generateBlockAtSlot(epoch4StartSlot);
+    importBlock(epoch4Block);
+
+    // Should now have justified epoch 3
+    assertThat(recentChainData.getJustifiedCheckpoint())
+        .map(Checkpoint::getEpoch)
+        .contains(UInt64.valueOf(3));
+
+    // The only block with the newly justified checkpoint is epoch4Block so it should become head
+    assertThat(recentChainData.getBestBlockRoot()).contains(epoch4Block.getRoot());
+  }
+
+  private void justifyEpoch(final ChainUpdater chainUpdater, final long epoch) {
+    final UInt64 nextEpochStartSlot = spec.computeStartSlotAtEpoch(UInt64.valueOf(epoch + 1));
+    // Advance chain to an epoch we can actually justify
+    prepEpochForJustification(chainUpdater, nextEpochStartSlot);
+
+    // Trigger epoch transition into next epoch.
+    importBlock(chainBuilder.generateBlockAtSlot(nextEpochStartSlot));
+
+    // Should now have justified epoch
+    assertThat(recentChainData.getJustifiedCheckpoint())
+        .map(Checkpoint::getEpoch)
+        .contains(UInt64.valueOf(epoch));
+  }
+
+  private void prepEpochForJustification(
+      final ChainUpdater chainUpdater, final UInt64 nextEpochStartSlot) {
+    UInt64 headSlot = chainUpdater.getHeadSlot();
+    final UInt64 targetHeadSlot = nextEpochStartSlot.minus(2);
+    while (headSlot.isLessThan(targetHeadSlot)) {
+      SignedBlockAndState headBlock = chainBuilder.generateBlockAtSlot(headSlot.plus(1));
+      importBlock(headBlock);
+      assertThat(recentChainData.getBestBlockRoot()).contains(headBlock.getRoot());
+      headSlot = headBlock.getSlot();
+    }
+
+    // Add a block with enough attestations to justify the epoch.
+    final BlockOptions epoch2BlockOptions = BlockOptions.create();
+    final UInt64 newBlockSlot = headSlot.increment();
+    chainBuilder
+        .streamValidAttestationsForBlockAtSlot(newBlockSlot)
+        .forEach(epoch2BlockOptions::addAttestation);
+    final SignedBlockAndState epoch2Block =
+        chainBuilder.generateBlockAtSlot(newBlockSlot, epoch2BlockOptions);
+
+    importBlock(epoch2Block);
   }
 
   @Test

--- a/protoarray/src/main/java/tech/pegasys/teku/protoarray/ForkChoiceStrategy.java
+++ b/protoarray/src/main/java/tech/pegasys/teku/protoarray/ForkChoiceStrategy.java
@@ -88,10 +88,15 @@ public class ForkChoiceStrategy implements BlockMetadataStore, ReadOnlyForkChoic
     return new ForkChoiceStrategy(protoArray, new ArrayList<>());
   }
 
-  public SlotAndBlockRoot findHead(final Checkpoint justifiedCheckpoint) {
+  public SlotAndBlockRoot findHead(
+      final Checkpoint justifiedCheckpoint, final Checkpoint finalizedCheckpoint) {
     protoArrayLock.readLock().lock();
     try {
-      final ProtoNode bestNode = protoArray.findHead(justifiedCheckpoint.getRoot());
+      final ProtoNode bestNode =
+          protoArray.findHead(
+              justifiedCheckpoint.getRoot(),
+              justifiedCheckpoint.getEpoch(),
+              finalizedCheckpoint.getEpoch());
       return new SlotAndBlockRoot(bestNode.getBlockSlot(), bestNode.getBlockRoot());
     } finally {
       protoArrayLock.readLock().unlock();
@@ -231,7 +236,7 @@ public class ForkChoiceStrategy implements BlockMetadataStore, ReadOnlyForkChoic
       protoArray.applyScoreChanges(deltas, justifiedEpoch, finalizedEpoch);
       balances = justifiedStateBalances;
 
-      return protoArray.findHead(justifiedRoot).getBlockRoot();
+      return protoArray.findHead(justifiedRoot, justifiedEpoch, finalizedEpoch).getBlockRoot();
     } finally {
       protoArrayLock.writeLock().unlock();
       votesLock.writeLock().unlock();

--- a/protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArray.java
+++ b/protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArray.java
@@ -144,7 +144,14 @@ public class ProtoArray {
    * @param justifiedRoot the root of the justified checkpoint
    * @return the best node according to fork choice
    */
-  public ProtoNode findHead(Bytes32 justifiedRoot) {
+  public ProtoNode findHead(Bytes32 justifiedRoot, UInt64 justifiedEpoch, UInt64 finalizedEpoch) {
+    if (!this.justifiedEpoch.equals(justifiedEpoch)
+        || !this.finalizedEpoch.equals(finalizedEpoch)) {
+      this.justifiedEpoch = justifiedEpoch;
+      this.finalizedEpoch = finalizedEpoch;
+      // Justified or finalized epoch changed we we have to re-evaluate all best descendants.
+      applyToNodes(this::updateBestDescendantOfParent);
+    }
     int justifiedIndex =
         indices
             .get(justifiedRoot)

--- a/protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java
+++ b/protoarray/src/test/java/tech/pegasys/teku/protoarray/ProtoArrayTest.java
@@ -90,11 +90,23 @@ class ProtoArrayTest {
     addBlock(1, blockRootB, Bytes32.ZERO);
 
     protoArray.applyProposerWeighting(proposerWeightingB);
-    assertThat(protoArray.findHead(GENESIS_CHECKPOINT.getRoot()).getBlockRoot())
+    assertThat(
+            protoArray
+                .findHead(
+                    GENESIS_CHECKPOINT.getRoot(),
+                    GENESIS_CHECKPOINT.getEpoch(),
+                    GENESIS_CHECKPOINT.getEpoch())
+                .getBlockRoot())
         .isEqualTo(blockRootB);
 
     protoArray.applyProposerWeighting(proposerWeightingA);
-    assertThat(protoArray.findHead(GENESIS_CHECKPOINT.getRoot()).getBlockRoot())
+    assertThat(
+            protoArray
+                .findHead(
+                    GENESIS_CHECKPOINT.getRoot(),
+                    GENESIS_CHECKPOINT.getEpoch(),
+                    GENESIS_CHECKPOINT.getEpoch())
+                .getBlockRoot())
         .isEqualTo(blockRootA);
   }
 

